src/routing/route.cc

```diff
@@ -190,31 +200,164 @@ LaneSRangeRelation Route::ComputeLaneSRangeRelation(const api::LaneSRange& lane_
-    if (lane_s_range_a.lane_id() != lane_s_range_b.lane_id()) {
-      return false;
-    }
-    const bool s1_is_in_range = lane_s_range_a.s_range().s1() + tolerance >= lane_s_range_b.s_range().s1() &&
-                                lane_s_range_a.s_range().s0() - tolerance <= lane_s_range_b.s_range().s1();
-    const bool s0_is_in_range = lane_s_range_a.s_range().s1() + tolerance >= lane_s_range_b.s_range().s0() &&
-                                lane_s_range_a.s_range().s0() - tolerance <= lane_s_range_b.s_range().s0();
-    return s1_is_in_range && s0_is_in_range;
+api::LaneSRoute Route::ComputeLaneSRoute(const api::RoadPosition& start_position) const {
+  MALIPUT_THROW_UNLESS(start_position.lane != nullptr);
+
+  const api::RoadPosition& end_road_position = end_route_position();
+  // Creates a point like api::LaneSRange to find the appropriate LaneSRangeIndex of the start
+  // and end positions.
+  const api::LaneSRange end_lane_s_range_point(end_road_position.lane->id(),
+                                               api::SRange(end_road_position.pos.s(), end_road_position.pos.s()));
+  const api::LaneSRange start_lane_s_range_point(start_position.lane->id(),
+                                                 api::SRange(start_position.pos.s(), start_position.pos.s()));
+
+  const std::optional<LaneSRangeIndex> start_lane_s_range_index = FindLaneSRangeIndex(start_lane_s_range_point);
+  MALIPUT_THROW_UNLESS(start_lane_s_range_index.has_value());
+
+  const std::optional<LaneSRangeIndex> end_lane_s_range_index = FindLaneSRangeIndex(end_lane_s_range_point);
+  MALIPUT_THROW_UNLESS(end_lane_s_range_index.has_value());
+
+  // Inserts @p lane_s_range_index into @p lane_s_range at the beginning.
+  auto prepend_lane_s_range_to_route = [this](const LaneSRangeIndex& lane_s_range_index,
+                                              std::vector<api::LaneSRange>* lane_s_ranges) -> void {
+    lane_s_ranges->insert(lane_s_ranges->begin(), GetLaneSRange(lane_s_range_index.first, lane_s_range_index.second));
   };
 
-  for (size_t i = 0; i < phases_.size(); ++i) {
-    for (size_t j = 0; j < phases_[i].lane_s_ranges().size(); ++j) {
-      if (is_lane_s_range_contained(phases_[i].lane_s_ranges()[j], lane_s_range)) {
-        return {std::make_pair(i, j)};
+  // Returns true when @p lane_s_range_index_a.first is equal to @p lane_s_range_index_b.first.
+  auto is_same_phase = [](const Route::LaneSRangeIndex& lane_s_range_index_a,
+                          const Route::LaneSRangeIndex& lane_s_range_index_b) -> bool {
+    return lane_s_range_index_a.first == lane_s_range_index_b.first;
+  };
+
+  LaneSRangeIndex current_lane_s_range_index = end_lane_s_range_index.value();
+  std::vector<api::LaneSRange> lane_s_ranges;
+
+  // The following double nested loop performs the routing within the Route from the ending api::LaneSRange
+  // towards the starting api::LaneSRange. The is always a possible logical solution, and it is based on
```

Comment:
"The is always" should be "It is always."

Reply:
Sorry about that. Thanks!